[MRG+1] Fix test of SingleInheritanceEstimator to not raise DeprecationWarning

zxcvbnius opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Fixes #8506 


#### What does this implement/fix? Explain your changes.
Fix test of SingleInheritanceEstimator to not raise DeprecationWarning. 
Set `sklearn.__version__` to the current version.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

zxcvbnius committed:
Fix test of SingleInheritanceEstimator to not raise DeprecationWarning

zxcvbnius committed:
Merge branch 'master' into sprint

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -436,6 +436,7 @@ def __init__(self, attribute_pickled=5):
 
     def __getstate__(self):
```
Remind me, @HolgerPeters, why this method does not call `super(...).__getstate__`...?

zxcvbnius replied:
Hi @jnothman thanks for your review!
I am a newbie here.
I think it may not be appropriate to use `super(...).__getstate__` because the next test case is `test_pickling_works_when_getstate_is_overwritten_in_the_child_class`.

lesteve replied:
@jnothman I think this was supposed to test the case where users have implemented an estimator that overrides `__getstate__`

HolgerPeters replied:
Yes, that is the intention of the test case. We had written our own serialization logic for estimators, and scikit-learn 0.18 broke all of our modelling code. For what its worth since the BaseEstimator does not contain parameters (and does not contain estimated parameters) scikit-learn shouldn't expect that users inheriting from `BaseEstimator` and overwriting `__getstate__` make that `super(..).__getstate__` call.

lesteve commented on sklearn/tests/test_base.py:
```diff
@@ -436,6 +436,7 @@ def __init__(self, attribute_pickled=5):
 
     def __getstate__(self):
         data = self.__dict__.copy()
+        data["_sklearn_version"] = sklearn.__version__
```
I am not entirely sure this is the right fix. I don't remember whether this test was supposed to test estimators outside scikit-learn or not. @HolgerPeters do you remember off the top of your head?

A closer look at #8324 may be needed.

HolgerPeters replied:
Not in this test case, [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_base.py#L412) we test estimators outside of sklearn.

HolgerPeters replied:
I still think that `BaseEstimator` shouldn't overwrite `__getstate__` and scikit-learn should probably just not try to do those version checks. I understand the intention of course, but basically scikit-learn user's must have a strategy in place for making sure that they don't unpickle estimators with the wrong scikit-learn version. To me seems scikit-learn tries here to address a general problem with `pickle` in a very scikit-learn specific way, which has unintended side-effects on user code that ran fine for years.

To really address the problem of serialization-versioning would probably mean to introduce special serialization code (which would mean A LOT of code writing). But then, you could start to maintain a versionizable serialization interface.

(@zxcvbnius this comment is not about your code in this PR, but about the background of the code you modify).

lesteve replied:
Thanks a lot for the details @HolgerPeters!

@zxcvbnius I think your change is fine. Add a comment on the previous line explaining that the line you added avoids a DeprecationWarning about unpickling an estimator from a different version.

HolgerPeters replied:
@lesteve I reworked my comment and [added it to the original ticket](https://github.com/scikit-learn/scikit-learn/issues/7135#issuecomment-285683380) because this PR is probably not the best place for this discussion.

HolgerPeters replied:
@zxcvbnius I also think the change is ok.

zxcvbnius replied:
@lesteve ok!
@HolgerPeters , really thanks for your detail explanation!

jnothman commented:
If the test is meant to reflect behaviour for an estimator outside of
scikit-learn, then it's the wrong fix. Ignoring the warning, or hacking the
__module__ would be better, with a comment, imo. This is just a side effect
is us having tests within the package, is it not?

On 11 Mar 2017 1:34 am, "Holger Peters" <notifications@github.com> wrote:

> *@HolgerPeters* commented on this pull request.
> ------------------------------
>
> In sklearn/tests/test_base.py
> <https://github.com/scikit-learn/scikit-learn/pull/8526#discussion_r105407630>
> :
>
> > @@ -436,6 +436,7 @@ def __init__(self, attribute_pickled=5):
>
>      def __getstate__(self):
>          data = self.__dict__.copy()
> +        data["_sklearn_version"] = sklearn.__version__
>
> @zxcvbnius <https://github.com/zxcvbnius> I also think the change is ok.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8526#discussion_r105407630>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xIHRgOwZHENxXnWeIAmDBlw5cWNks5rkV-HgaJpZM4MTMLu>
> .
>

lesteve commented:
> If the test is meant to reflect behaviour for an estimator outside of scikit-learn, then it's the wrong fix

According to https://github.com/scikit-learn/scikit-learn/pull/8526#discussion_r105400747, it looks like this is not the purpose of the test.

jnothman commented:
I still suspect ignore_warnings is a more appropriate solution

lesteve commented:
> I still suspect ignore_warnings is a more appropriate solution

Fine with me. @zxcvbnius can add a @ignore_warnings to the test function please?

zxcvbnius committed:
[MRG+1] Fix test of SingleInheritanceEstimator to not raise DeprecationWarning (#8526)

Fix: test of SingleInheritanceEstimator to not raise DeprecationWarning (#8526)

- Ignore a DeprecationWarning about unpickling an estimator from a different version

jnothman commented on sklearn/tests/test_base.py:
```diff
@@ -436,11 +437,11 @@ def __init__(self, attribute_pickled=5):
         return data
 
 
+@ignore_warnings(category=(DeprecationWarning, UserWarning))
```
why DeprecationWarning?

zxcvbnius replied:
I think I should not be confused by `DeprecationWarning` in the title.
According to the following error message:
**UserWarning: Trying to unpickle estimator SingleInheritanceEstimator from version pre-0.18 when using version 0.19.dev0. This might lead to breaking code or invalid results. Use at your own risk.**
We just need `UserWarning`

zxcvbnius committed:
[MRG+2] Fix test of SingleInheritanceEstimator to not raise DeprecationWarning (#8526)

Fix: test of SingleInheritanceEstimator to not raise DeprecationWarning (#8526)

  - Test of SingleInheritanceEstimator would raise UserWarning, not DeprecationWarning
  - Ignore a UserWarning about unpickling an estimator from a different version

jnothman commented:
LGTM (I ran the tests).

PS: you are using very strange commit messages. They should describe what change the commit makes. MRG+1, MRG+2 should have nothing to do with your commit message.

zxcvbnius commented:
@jnothman Ohh yes, you're right. Thanks for your reminder. I will be careful next time.

MechCoder commented:
Thanks!
